### PR TITLE
Fix floating reads

### PR DESCRIPTION
**Pull request by gigiblender**
#### Description

This PR fixes the following:
- floating reads scheduled in the wrong block are now replaced by fixed reads in the low tier (FixReadsPhase)
- implement remainder operation for the PTX backend
- fix loop labels for loop back edges for the PTX backend

#### Problem description

Graal will lower memory reads to floating read nodes in order to perform better optimisations. However, some floating read nodes might be scheduled in the wrong Basic Block. Adding the `FixReadsPhase` will find such nodes and replace them with fixed read nodes.

Also, loop labels for loop back edges were broken in the PTX backend. We were previously adding the loop label at the loop condition, but there might be dataflow attached to the loop header that was not executed.

#### Backend/s tested

- [X] OpenCL
- [X] PTX

#### OS tested

- [X] Linux
- [ ] OSx
- [ ] Windows

#### Did you check on FPGAs?

If possible, check your changes on FPGAs.

- [ ] Yes
- [X] No

#### How to test the new patch?

Provide instructions about how to test the new patch. 

Running the unittests should cover everything.

### CODE REVIEW

File: unittests/src/main/java/uk/ac/manchester/tornado/unittests/common/OpenCLNotSupported.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package uk.ac.manchester.tornado.unittests.common;
```
Comment (jjfumero) [inline unittests/src/main/java/uk/ac/manchester/tornado/unittests/common/OpenCLNotSupported.java:1]: Add license - Apache2 header

File: unittests/src/main/java/uk/ac/manchester/tornado/unittests/common/OpenCLNotSupported.java [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+public class OpenCLNotSupported extends RuntimeException {
```
Comment (jjfumero) [inline unittests/src/main/java/uk/ac/manchester/tornado/unittests/common/OpenCLNotSupported.java:3]: Rename this class `OpenCLNotSupported` -> `TornadoOpenCLNotSupported`

File: unittests/src/main/java/uk/ac/manchester/tornado/unittests/common/TornadoTestBase.java [hunk 3]
```diff
@@ -93,4 +93,11 @@ public void checkForPTX() {
         }
     }
 
+    public void checkForOpenCL() {
+        int driverIndex = TornadoRuntime.getTornadoRuntime().getDefaultDevice().getDriverIndex();
+        if ("OpenCL".equals(TornadoRuntime.getTornadoRuntime().getDriver(driverIndex).getName())) {
```
Comment (jjfumero) [inline unittests/src/main/java/uk/ac/manchester/tornado/unittests/common/TornadoTestBase.java:98]: I suggest adding the method `getBackendType`  

`TornadoRuntime.getTornadoRuntime().getBackendType(driverIndex)`

This returns a value based on an enum:

`OCL, PTX, etc.` 

And we avoid String comparisons.

File: unittests/src/main/java/uk/ac/manchester/tornado/unittests/tools/TornadoHelper.java [hunk 4]
```diff
@@ -162,13 +163,20 @@ static void runTestVerbose(String klassName, String methodName) throws ClassNotF
                     continue;
                 }
 
-                if (result.getFailures().stream().filter(e -> (e.getException() instanceof PTXNotSupported)).count() > 0) {
+                if (result.getFailures().stream().anyMatch(e -> (e.getException() instanceof PTXNotSupported))) {
```
Comment (jjfumero) [inline unittests/src/main/java/uk/ac/manchester/tornado/unittests/tools/TornadoHelper.java:166]: Let's also update the `PTXNotSupported` -> `TornadoPTXNotSupported`

**Review by jjfumero — CHANGES_REQUESTED** [review]

File: unittests/src/main/java/uk/ac/manchester/tornado/unittests/common/TornadoOpenCLNotSupported.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package uk.ac.manchester.tornado.unittests.common;
+
+public class TornadoOpenCLNotSupported extends RuntimeException {
```
Comment (stratika) [inline unittests/src/main/java/uk/ac/manchester/tornado/unittests/common/TornadoOpenCLNotSupported.java:21]: Refactoring: Tornado -> TornadoVM

File: unittests/src/main/java/uk/ac/manchester/tornado/unittests/common/TornadoPTXNotSupported.java [hunk 6]
```diff
@@ -18,13 +18,13 @@
 
 package uk.ac.manchester.tornado.unittests.common;
 
-public class PTXNotSupported extends RuntimeException {
+public class TornadoPTXNotSupported extends RuntimeException {
```
Comment (stratika) [inline unittests/src/main/java/uk/ac/manchester/tornado/unittests/common/TornadoPTXNotSupported.java:21]: Refactoring: TornadoPTXNotSupported -> TornadoVMPTXNotSupported

File: unittests/src/main/java/uk/ac/manchester/tornado/unittests/loops/TestLoops.java [hunk 7]
```diff
@@ -827,4 +829,58 @@ public void testReverseOrderLoops() {
...
+        }
+    }
+
+    private static int getNumber(int[] in, int base, int offset, int multiplier) {
+        // Perform some address computation
+        return in[base * multiplier + offset];
+    }
+
+    /**
+     * Make sure that on the final iteration of the loop, the address computed in {@link #getNumber} has been updated
```
Comment (stratika) [inline unittests/src/main/java/uk/ac/manchester/tornado/unittests/loops/TestLoops.java:853]: on the last iteration -> at the last iteration

File: unittests/src/main/java/uk/ac/manchester/tornado/unittests/loops/TestLoops.java [hunk 8]
```diff
@@ -827,4 +829,58 @@ public void testReverseOrderLoops() {
...
+     * with the latest value of the induction variable, in order to be used in the if condition.
+     *
+     * TODO this test currently fails on the OpenCL backend because we do not support complex data flow attached to the loop begin.
```
Comment (stratika) [inline unittests/src/main/java/uk/ac/manchester/tornado/unittests/loops/TestLoops.java:856]: we do not support -> TornadoVM does not support

**Review by stratika — CHANGES_REQUESTED** [review]

File: unittests/src/main/java/uk/ac/manchester/tornado/unittests/tools/TornadoHelper.java (hunk 4 above)
Comment (stratika) [reply unittests/src/main/java/uk/ac/manchester/tornado/unittests/tools/TornadoHelper.java]: I think it is better to change `Tornado` with `TornadoVM` for every new reference that we add. What do you think @jjfumero?

**Review by stratika — APPROVED** [review]

**Review by jjfumero — APPROVED** [review]

**Pull request merged**